### PR TITLE
Add Cocos2d-x feature guides

### PR DESCRIPTION
**Pull request by Sporiff**
Closes https://adjustcom.atlassian.net/browse/ADAPP-18657

### CODE REVIEW

**Comment by Sporiff** [comment]
@nonelse @uerceg We're still adding a few bits and pieces, but this is mostly ready for review here: https://dev-docs-git-adapp-18657-cocos-2-dx-v-5-201a97-product-content.vercel.app/en/sdk/cocos2dx

File: src/content/docs/sdk/cocos2dx/v5/features/device-info.mdoc [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+});
+```
+
+## Amazon Advertiser ID {% #amazon-adid %}
+
+The Amazon Advertising ID (Amazon Ad ID) is a device-specific identifier for Android devices. Call the `getAmazonAdId` method to asynchronously return the Amazon Ad ID as a `string` to a delegate function.
+
+```cpp
+Adjust2dx::getAmazonAdId([](std::string adId) {
+    std::cout << "\nAmazon Ad ID = " << adId;
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/device-info.mdoc:67]:
```suggestion
    std::cout << "\nAmazon Ad ID = " << amazonAdId;
```

File: src/content/docs/sdk/cocos2dx/v5/features/skan.mdoc [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+## Update conversion values {% #update-conversion-values %}
+
+Conversion values are a mechanism used to track user behavior in SKAN. You can map 64 conditions to values from 0 through 63 and send this integer value to SKAN on user install. This gives you insight into how your users interact with your app in the first few days.
+
+If you manage your conversion values with Adjust, the servers update this value in the SDK. You can also update this value by using the `updateSkanConversionValue` method. This method wraps [Apple's `updatePostbackConversionValue` method](https://developer.apple.com/documentation/storekit/skadnetwork/4097267-updatepostbackconversionvalue). You MUST pass the following arguments:
+
+- `fineValue` (`int`): Your conversion value. Must be between `0` and `63`.
+- `coarseValue` (`string` [`SKAdNetwork.CoarseConversionValue`](https://developer.apple.com/documentation/storekit/skadnetwork/coarseconversionvalue)): The coarse conversion value. This value is used if your app doesn't have sufficient installs to reach the privacy threshold. The following values are allowed:
+  - `low` (`SKAdNetworkCoarseConversionValueLow`)
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/skan.mdoc:39]:
```suggestion
  - `"low"` (`SKAdNetworkCoarseConversionValueLow`)
```

File: src/content/docs/sdk/cocos2dx/v5/features/skan.mdoc [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  - `medium` (`SKAdNetworkCoarseConversionValueMedium`)
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/skan.mdoc:40]:
```suggestion
  - `"medium"` (`SKAdNetworkCoarseConversionValueMedium`)
```

File: src/content/docs/sdk/cocos2dx/v5/features/skan.mdoc [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  - `high` (`SKAdNetworkCoarseConversionValueHigh`)
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/skan.mdoc:41]:
```suggestion
  - `"high"` (`SKAdNetworkCoarseConversionValueHigh`)
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+{% tabs %}
+{% tab title="App Store" sync="appstore" %}
+
+Instantiate an `AdjustAppStoreSubscription2dx` object with the following arguments:
+
+| Argument      | Type | Description                                                                                                                                                                                                                                                                                         |
+| --------------- | --------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `price`         | `string`    | The price of the subscription                                                                                                                                                                                                                                                                       |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:40]:
```suggestion
| `price`         | `std::string`    | The price of the subscription                                                                                                                                                                                                                                                                       |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+| `currency`      | `string`    | The currency of the subscription. Formatted as the [`currencyCode`](https://developer.apple.com/documentation/foundation/nslocale/1642836-currencycode?language=objc) of the [`priceLocale`](https://developer.apple.com/documentation/storekit/skproduct/1506145-pricelocale?language=objc) object |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:41]:
```suggestion
| `currency`      | `std::string`    | The currency of the subscription. Formatted as the [`currencyCode`](https://developer.apple.com/documentation/foundation/nslocale/1642836-currencycode?language=objc) of the [`priceLocale`](https://developer.apple.com/documentation/storekit/skproduct/1506145-pricelocale?language=objc) object |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+| `transactionId` | `string`    | Your ID for the transaction                                                                                                                                                                                                                                                                         |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:42]:
```suggestion
| `transactionId` | `std::string`    | Your ID for the transaction                                                                                                                                                                                                                                                                         |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+```
+
+{% /tab %}
+{% tab title="Play Store" sync="playstore" %}
+
+Instantiate an `AdjustPlayStoreSubscription2dx` object with the following arguments:
+
+| Argument       | Type | Description                                                                                                                                                                                 |
+| --------------- | --------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `price`         | `double`    | The price of the subscription                                                                                                                                                               |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:55]:
```suggestion
| `price`         | `std::string`    | The price of the subscription                                                                                                                                                               |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+| `currency`      | `string`    | The currency of the subscription                                                                                                                                                            |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:56]:
```suggestion
| `currency`      | `std::string`    | The currency of the subscription                                                                                                                                                            |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+| `sku`           | `string`    | The ID of the product                                                                                                                                                                       |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:57]:
```suggestion
| `sku`           | `std::string`    | The ID of the product                                                                                                                                                                       |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+| `orderId`       | `string`    | Your ID for the transaction                                                                                                                                                                 |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:58]:
```suggestion
| `orderId`       | `std::string`    | Your ID for the transaction                                                                                                                                                                 |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+| `signature`     | `string`    | The signature of the purchase data                                                                                                                                                          |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:59]:
```suggestion
| `signature`     | `std::string`    | The signature of the purchase data                                                                                                                                                          |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+| `purchaseToken` | `string`    | The unique token of the transaction. See [Google's documentation](<https://developer.android.com/reference/com/android/billingclient/api/Purchase#getPurchaseToken()>) for more information |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:60]:
```suggestion
| `purchaseToken` | `std::string`    | The unique token of the transaction. See [Google's documentation](<https://developer.android.com/reference/com/android/billingclient/api/Purchase#getPurchaseToken()>) for more information |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+You can record the date on which the user purchased a subscription. The SDK returns this data for you to report on.
+
+{% tabs %}
+{% tab title="App Store" sync="appstore" %}
+
+Call the `setTransactionDate` method on your subscription object to record the timestamp of the subscription.
+
+| Argument          | Type      | Description                              |
+| ----------------- | --------- | ---------------------------------------- |
+| `transactionDate` | `string`  | The timestamp of the subscription        |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:79]:
```suggestion
| `transactionDate` | `std::string`  | The timestamp of the subscription        |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+```
+
+{% /tab %}
+{% tab title="Play Store" sync="playstore" %}
+
+Call the `setPurchaseTime` method on your subscription object to record the timestamp of the subscription.
+
+| Argument          | Type      | Description                              |
+| ----------------- | --------- | ---------------------------------------- |
+| `purchaseTime`    | `string`  | The timestamp of the subscription        |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:93]:
```suggestion
| `purchaseTime`    | `std::string`  | The timestamp of the subscription        |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+{% /tab %}
+{% /tabs %}
+
+### Record the purchase region (iOS only) {% #record-the-purchase-region-ios-only %}
+
+You can record the region in which the user purchased a subscription. To do this, call the `setSalesRegion` method on your subscription object and pass the country code as a **string**. This needs to be formatted as the [`countryCode`](https://developer.apple.com/documentation/storekit/storefront/3792000-countrycode) of the [`Storefront`](https://developer.apple.com/documentation/storekit/storefront) object.
+
+| Argument          | Type      | Description                              |
+| ----------------- | --------- | ---------------------------------------- |
+| `salesRegion`     | `string`  | The country code for the subscription    |
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:108]:
```suggestion
| `salesRegion`     | `std::string`  | The country code for the subscription    |
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+   - The partner parameters are set to _"key", "value"_ and _"foo", "bar"_.
+1. At the end of your function, send the information to Adjust by calling `trackAppStoreSubscription` with your `AdjustAppStoreSubscription2dx` instance as an argument.
+
+```cpp
+#include "Adjust/Adjust2dx.h"
+
+std::string appToken = "{% $variables.config.token %}";
+std::string environment = AdjustEnvironmentSandbox2dx;
+
+AdjustConfig2dx adjustConfig = AdjustConfig2dx(appToken, environment, false);
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:191]:
```suggestion
AdjustConfig2dx adjustConfig = AdjustConfig2dx(appToken, environment);
```

File: src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc [hunk 18]
```diff
@@ -0,0 +1,232 @@
...
+   - The partner parameters are set to _"key", "value"_ and _"foo", "bar"_.
+1. At the end of your function, send the information to Adjust by calling `trackPlayStoreSubscription` with your `AdjustAppStoreSubscription2dx` instance as an argument.
+
+```cpp
+#include "Adjust/Adjust2dx.h"
+
+std::string appToken = "{% $variables.config.token %}";
+std::string environment = AdjustEnvironmentSandbox2dx;
+
+AdjustConfig2dx adjustConfig = AdjustConfig2dx(appToken, environment, false);
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/subscriptions.mdoc:219]:
```suggestion
AdjustConfig2dx adjustConfig = AdjustConfig2dx(appToken, environment);
```

File: src/content/docs/sdk/cocos2dx/v5/configuration.mdoc [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+To instantiate your config object, create a new `AdjustConfig2dx` instance and pass the following parameters:
+
+- `appToken` (`String`): Your [Adjust app token](https://help.adjust.com/en/article/app-token-and-reporting-currency#view-your-app-details).
+- `environment` (`String`): The environment you want to run the SDK in. Pass `AdjustEnvironmentSandbox2dx` to run the SDK in sandbox mode for testing. Pass `AdjustEnvironmentProduction2dx` to run the SDK in production mode for release.
+- `allowSuppressLogLevel` (`Boolean`): Whether to suppress all logging. Set to `true` to suppress logging or `false` to enable logging.
+
+```cpp
+#include "Adjust/Adjust2dx.h"
+
+std::string appToken = "{% $variables.config.token %}";
```
[the remark this answers is not in the document]
Comment (Sporiff) [reply src/content/docs/sdk/cocos2dx/v5/configuration.mdoc:32]: @uerceg I've updated the token to `{YOUR_APP_TOKEN}` in all cases.

**Comment by nonelse** [comment]
I'm reviewing, by looking at other targets Unity specifically, besides at the code changes, so some of the questions might not apply to cocos 2dx and @uerceg can probably clarify if it makes sense for Cocos 2dx or not. Starting with the landing page for Cocos 2dx:
Android does not contain App Set Identifier like Unity, should it?

**Comment by nonelse** [comment]
Install referrer contains more/different explanation than Unity, does it need to?

**Comment by nonelse** [comment]
Does not contain Meta referrer integration, should it?

**Comment by nonelse** [comment]
Does not contain an App Tracking Transparency consent dialog section in iOs, should it?

**Comment by nonelse** [comment]
Does not contain an Deep linking section in iOs, should it?

File: src/content/docs/sdk/cocos2dx/v5/configuration.mdoc [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+
+## Instantiate your config object {% #instantiate-your-config-object %}
+
+To configure the Adjust SDK, you need to instantiate an `AdjustConfig2dx` object. This object contains the **read-only** configuration options that you need to pass to the Adjust SDK.
+
+To instantiate your config object, create a new `AdjustConfig2dx` instance and pass the following parameters:
+
+- `appToken` (`std::string`): Your [Adjust app token](https://help.adjust.com/en/article/app-token-and-reporting-currency#view-your-app-details).
+- `environment` (`std::string`): The environment you want to run the SDK in. Pass `AdjustEnvironmentSandbox2dx` to run the SDK in sandbox mode for testing. Pass `AdjustEnvironmentProduction2dx` to run the SDK in production mode for release.
+- `allowSuppressLogLevel` (`bool`): Whether to suppress all logging. Set to `true` to suppress logging or `false` to enable logging.
```
Comment (nonelse) [inline src/content/docs/sdk/cocos2dx/v5/configuration.mdoc:27]: only the constructor with `allowSuppressLogLevel` is mentioned. Could it be referred as optional and that `false` would be used with the shorter constructor? (The constructor without allowSuppressLogLevel is used later in examples)

File: src/content/docs/sdk/cocos2dx/v5/features/ad-revenue.mdoc [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+
+#### Ad revenue network
+
+To send the network associated with ad revenue, call the `setAdRevenueNetwork` method of your `AdjustAdRevenue2dx` instance with the following argument:
+
+- `adRevenueNetwork` (`std::string`): The name of the network associated with the ad revenue.
+
+```cpp
+AdjustAdRevenue2dx adjustAdRevenue = AdjustAdRevenue2dx("{% $variables.adRevenue.source %}");
+adjustAdRevenue.adRevenueNetwork("{% $variables.adRevenue.adRevenueNetwork %}");
```
Comment (nonelse) [inline src/content/docs/sdk/cocos2dx/v5/features/ad-revenue.mdoc:81]: should be `adjustAdRevenue.setAdRevenueNetwork(std::string)`

File: src/content/docs/sdk/cocos2dx/v5/features/ad-revenue.mdoc [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+
+#### Ad revenue unit
+
+To send the advertising unit that earned the revenue, call the `setAdRevenueUnit` method of your `AdjustAdRevenue2dx instance with the following argument:
+
+- `adRevenueUnit` (`std::string`): The name of the ad unit associated with the ad revenue.
+
+```cpp
+AdjustAdRevenue2dx adjustAdRevenue = AdjustAdRevenue2dx("{% $variables.adRevenue.source %}");
+adjustAdRevenue.adRevenueUnit("{% $variables.adRevenue.adRevenueUnit %}");
```
Comment (nonelse) [inline src/content/docs/sdk/cocos2dx/v5/features/ad-revenue.mdoc:93]: should be `setAdRevenueUnit(std::string)`

**Comment by uerceg** [comment]
@nonelse, I'll just reply to those I think I know the answer to:

- ATT dialog, it's here: https://dev-docs-git-adapp-18657-cocos-2-dx-v-5-201a97-product-content.vercel.app/en/sdk/cocos2dx/features/att
- Deeplinking in iOS, [the chapter](https://dev-docs-git-adapp-18657-cocos-2-dx-v-5-201a97-product-content.vercel.app/en/sdk/cocos2dx/features/deep-links) in general just points out to the native guides on how to set things up. Is that enough or did you have something else in mind?
- Meta install referrer, nice catch, I haven't noticed it. 👍 /cc @Sporiff
- App set identifier, nice catch, I haven't noticed this one either. 👍 /cc @Sporiff

File: src/content/docs/sdk/cocos2dx/v5/configuration.mdoc (hunk 20 above)
Comment (Sporiff) [reply src/content/docs/sdk/cocos2dx/v5/configuration.mdoc:27]: @nonelse I'll update this

**Comment by Sporiff** [comment]
@uerceg @nonelse I'm going to keep this changeset limited to Cocos. If we need to change something in the other platforms, we'll do that separately.

For the Meta install referrer, do we bundle this like in Unity so that the reader just has to configure their AndroidManifest.xml and their FbAppid? Or is there a plugin for this?

**Comment by uerceg** [comment]
> For the Meta install referrer, do we bundle this like in Unity so that the reader just has to configure their AndroidManifest.xml and their FbAppid? Or is there a plugin for this?

@Sporiff Handling of META install referrer is as of v5 extracted to naive plugin which needs to be added to any platform in case clients want to use this feature. Same with Cocos2d-x. And similar to other non-native SDKs, we have:

```cpp
setFbAppId(std::string fbAppId);
```

in the `AdjustConfig2dx` to set the ID.

**Comment by Sporiff** [comment]
@uerceg So they need to be linked to the Android Meta Install Referrer plugin guide to install it in the Android project, then configure their `FbAppid`, is that right?

File: src/content/docs/sdk/cocos2dx/v5/configuration.mdoc [hunk 23]
```diff
@@ -0,0 +1,221 @@
...
+
+Use the methods in this document to configure the behavior of the Adjust SDK.
+
+## Instantiate your config object {% #instantiate-your-config-object %}
+
+To configure the Adjust SDK, you need to instantiate an `AdjustConfig2dx` object. This object contains the **read-only** configuration options that you need to pass to the Adjust SDK.
+
+To instantiate your config object, create a new `AdjustConfig2dx` instance and pass the following parameters:
+
+{% deflist %}
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/configuration.mdoc:25]: Screenshot 2024-09-30 at 14 50 30
Is this how the preview should look like?

File: src/content/docs/sdk/cocos2dx/v5/features/events.mdoc [hunk 24]
```diff
@@ -0,0 +1,244 @@
...
+```cpp
+#include "Adjust/Adjust2dx.h"
+
+std::string appToken = "{% $variables.config.token %}";
+std::string environment = AdjustEnvironmentSandbox2dx;
+
+AdjustConfig2dx adjustConfig = AdjustConfig2dx(appToken, environment);
+Adjust2dx::initSdk(adjustConfig);
+
+AdjustEvent2dx adjustEvent = new AdjustEvent2dx("g3mfiw");
```
Comment (uerceg) [inline src/content/docs/sdk/cocos2dx/v5/features/events.mdoc:180]:
```suggestion
AdjustEvent2dx adjustEvent = AdjustEvent2dx("g3mfiw");
```

File: src/content/docs/sdk/cocos2dx/v5/configuration.mdoc (hunk 23 above)
Comment (Sporiff) [reply src/content/docs/sdk/cocos2dx/v5/configuration.mdoc:25]: @uerceg Should be fixed

**Comment by nonelse** [comment]
Continuing with the Unity comparison, I noticed that there are 2 feature sections that are not present in the Cocos2dx v5: `Send preinstalled app activity` and `Purchase verification`. Just wanted to check if they are meant to be added (or not)

**Comment by Sporiff** [comment]
@nonelse I've added the guides you mentioned. Please check them and verify the code examples.

**Review by uerceg — APPROVED** [review]
:shipit:

**Pull request merged**